nearswap/src/twap.rs

```diff
@@ -0,0 +1,601 @@
+        let current_o = &self.observations.get(self.current_idx).unwrap();
+        
+        let total_observe;
+        let price1cumu;
+        let price2cumu;
+        if left_index == 0 {
+            total_observe = current_o.num_of_observations;
+            price1cumu = current_o.price1_cumulative;
+            price2cumu = current_o.price2_cumulative;
+        } else {
+            let prev_o = &self.observations.get(left_index - 1).unwrap();
+            total_observe = current_o.num_of_observations
+                                    - prev_o.num_of_observations;
+            price1cumu = current_o.price1_cumulative
+                                    - prev_o.price1_cumulative;
+            price2cumu = current_o.price2_cumulative
+                                    - prev_o.price2_cumulative;
+        }
+        let mean1 = price1cumu / total_observe;
+        let mean2 = price2cumu / total_observe;
+        return (mean1, mean2);
+    }
+
+    pub(crate) fn log_observation(&mut self, timestamp: u64, price1: u128, price2: u128) {
+        // update current index
+        if self.observations.len() == 0 {
+            self.current_idx = self.initialize(timestamp, price1, price2);
+        } else {
+            self.current_idx = self.write(
+                timestamp,
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        self.update_mean(self.observations.len());
+    }
+
+    pub fn update_mean(&mut self, len: u64) {
+        self.mean_1min = self.calculate_mean(Mean::M_1MIN, len);
+        self.mean_5min = self.calculate_mean(Mean::M_5MIN, len);
+        self.mean_1h = self.calculate_mean(Mean::M_1H, len);
+        self.mean_12h = self.calculate_mean(Mean::M_12H, len);
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Observation {
+    // timestamp
+    pub block_timestamp: u64,
+    // Number of observations till block_timestamp
+    pub num_of_observations: u128,
+    // cumulative price of token1 till block_timestamp
+    pub price1_cumulative: u128,
+    // cumulative price of token2 till block_timestamp
+    pub price2_cumulative: u128,
+}
+
+/**
+Transforms a previous observation into a new observation.
+Parameters:
++ `block_timestamp`: _must_ be chronologically equal to or greater than last.block_timestamp.
++ `last`: The specified observation to be transformed.
++ `price1`: price of first token.
++ `price2`: price of second token.
+*/
+pub fn transform(
+    last: &Observation,
+    block_timestamp: u64,
+    price1: u128,
+    price2: u128
+) -> Observation {
+    return
+        Observation {
+            block_timestamp: block_timestamp,
+            num_of_observations: last.num_of_observations + 1,
+            price1_cumulative: last.price1_cumulative + price1,
+            price2_cumulative: last.price2_cumulative + price2,
+        };
+}
```

Comment:
so, why can't we make it a method of the `Observation` class?
```
pub fn transform(&self, ...) -> Observation {... }
```

Reply:
previously it was added in Observation class, but it was updated because of the comment. maybe i misunderstood

Reply:
moved to observation